File: Semester 2/Algoritmica Grafelor/Laboratory/Lab 1/C++/graph.cpp
#include "graph.h"

VertexError::VertexError(const char* c) {
    msg = c;
}

const char *VertexError::what() const noexcept {
    return msg;
}

EdgeError::EdgeError(const char *c) {
    msg = c;
}

const char *EdgeError::what() const noexcept {
    return msg;
}

Graph::Graph() = default;

Graph::Graph(int vertices_no) {
    if (vertices_no < 0)
        throw VertexError("Invalid number of vertices.");
    for (int i = 0; i < vertices_no; ++i)
        add_vertex(i);
}

Graph::Graph(int vertices_no, int edges_no) {
    if (edges_no < 0 || edges_no > vertices_no*(vertices_no+1)/2)
        throw EdgeError("Invalid number of edges.");
    if (vertices_no < 0)
        throw VertexError("Invalid number of vertices.");
    for (int i = 0; i < vertices_no; ++i)
        add_vertex(i);
    std::mt19937 rng(std::chrono::steady_clock::now().time_since_epoch().count());
    for (int i = 0; i < edges_no; ++i){
        int a, b, c;
        do{
            a = int(rng()%2147483647)%vertices_no;
            b = int(rng()%2147483647)%vertices_no;
            c = int(rng()%2147483647)%1000000;
        }while(is_edge(a, b));
        add_edge(a, b, c);
    }
}

set_int::const_iterator Graph::vertices_begin() {
    return vertices.begin();
}

set_int::const_iterator Graph::vertices_end() {
    return vertices.end();
}

set_int::const_iterator Graph::neighbours_begin(int vertex) {
    if (!is_vertex(vertex))
        throw VertexError("Invalid vertex.");
    return edges[vertex].begin();
}

set_int::const_iterator Graph::neighbours_end(int vertex) {
    if (!is_vertex(vertex))
        throw VertexError("Invalid vertex.");
    return edges[vertex].end();
}

set_int::const_iterator Graph::transpose_begin(int vertex) {
    if (!is_vertex(vertex))
        throw VertexError("Invalid vertex.");
    return transpose[vertex].begin();
}

set_int::const_iterator Graph::transpose_end(int vertex) {
    if (!is_vertex(vertex))
        throw VertexError("Invalid vertex.");
    return transpose[vertex].end();
}

bool Graph::is_edge(int vertex1, int vertex2) {
    return edges.find(vertex1) != edges.end() && edges[vertex1].find(vertex2) != edges[vertex1].end();
}

bool Graph::is_edge(pair_int edge) {
    return edges.find(edge.first) != edges.end() && edges[edge.first].find(edge.second) != edges[edge.first].end();
}

bool Graph::is_vertex(int vertex) {
    return vertices.find(vertex) != vertices.end();
}

int Graph::count_vertices(){
    return vertices.size();
}

int Graph::in_degree(int vertex) {
    if (transpose.find(vertex) != transpose.end())
        return transpose[vertex].size();
    throw VertexError("The specified vertex does not exist.");
}

int Graph::out_degree(int vertex) {
    if (edges.find(vertex) != edges.end())
        return edges[vertex].size();
    throw VertexError("The specified vertex does not exist.");
}

int Graph::get_edge_cost(int vertex1, int vertex2) {
    auto it = cost.find(Edge(vertex1, vertex2, 0));
    if (it == cost.end())
        throw EdgeError("The specified edge does not exist.");
    return it->cost;
}

void Graph::set_edge_cost(int vertex1, int vertex2, int new_cost) {
    Edge new_edge(vertex1, vertex2, new_cost);
    auto it = cost.find(new_edge);
    if (it == cost.end())
        throw EdgeError("The specified edge does not exist.");
    cost.erase(new_edge);
    cost.insert(new_edge);
}

void Graph::add_vertex(int vertex) {
    if (is_vertex(vertex))
        throw VertexError("Cannot add a vertex which already exists.");
    vertices.insert(vertex);
    edges[vertex] = set_int();
    transpose[vertex] = set_int();
}

void Graph::remove_vertex(int vertex) {
    if (!is_vertex(vertex))
        throw VertexError("Cannot remove a vertex which doesn't exist.");
    std::vector<int> to_remove;
    for (auto node : edges[vertex])
        to_remove.emplace_back(node);
    for (auto node : to_remove)
        remove_edge(vertex, node);
    to_remove.clear();
    for (auto node : transpose[vertex])
        to_remove.emplace_back(node);
    for (auto node : to_remove)
        remove_edge(node, vertex);
    edges.erase(vertex);
    transpose.erase(vertex);
    vertices.erase(vertices.find(vertex));
}

void Graph::add_edge(int vertex1, int vertex2, int edge_cost) {
    if (is_edge(std::make_pair(vertex1, vertex2)))
        throw EdgeError("The specified edge already exists");
    if (!is_vertex(vertex1) || !is_vertex(vertex2))
        throw EdgeError("The vertices on the edge do not exist.");
    edges[vertex1].insert(vertex2);
    transpose[vertex2].insert(vertex1);
    cost.insert(Edge(vertex1, vertex2, edge_cost));
}

void Graph::remove_edge(int vertex1, int vertex2) {
    if (!is_edge(std::make_pair(vertex1, vertex2)))
        throw EdgeError("The specified edge does not exist.");
    cost.erase(Edge(vertex1, vertex2, 0));
    edges[vertex1].erase(edges[vertex1].find(vertex2));
    transpose[vertex2].erase(transpose[vertex2].find(vertex1));
}

int Graph::count_edges() {
    return cost.size();
}

set_edge::const_iterator Graph::edges_begin() {
    return cost.begin();
}

set_edge::const_iterator Graph::edges_end() {
    return cost.end();
}

Graph read_file(const char* file_path) {
    std::ifstream fin(file_path);
    int n, m;
    fin >> n >> m;
    Graph g(n);
    for (int i = 1; i <= m; ++i){
        int a, b, c;
        fin >> a >> b >> c;
        g.add_edge(a, b, c);
    }
    fin.close();
    return g;
}

void write_file(const char* file_path, Graph &graph) {
    std::ofstream fout(file_path);
    fout << graph.count_vertices() << " " << graph.count_edges() << "\n";
    for (auto it = graph.vertices_begin(); it != graph.vertices_end(); ++it)
        for (auto node = graph.neighbours_begin(*it); node != graph.neighbours_end(*it); ++node)
            fout << *it << " " << *node << " " << graph.get_edge_cost(*it, *node) << "\n";
    fout.close();
}

Graph random_graph(int vertices_no, int edges_no)
{
    Graph g;
    if (edges_no < 0 || edges_no > vertices_no*(vertices_no+1)/2)
        throw EdgeError("Invalid number of edges.");
    if (vertices_no < 0)
        throw VertexError("Invalid number of vertices.");
    for (int i = 0; i < vertices_no; ++i)
        g.add_vertex(i);
    std::mt19937 rng(std::chrono::steady_clock::now().time_since_epoch().count());
    for (int i = 0; i < edges_no; ++i){
        int a, b, c;
        do{
            a = int(rng()%2147483647)%vertices_no;
            b = int(rng()%2147483647)%vertices_no;
            c = int(rng()%2147483647)%1000000;
        }while(g.is_edge(a, b));
        g.add_edge(a, b, c);
    }
    return g;
}

Edge::Edge(int _vertex1, int _vertex2, int _cost) {
    vertex1 = _vertex1;
    vertex2 = _vertex2;
    cost = _cost;
}

bool Edge::operator==(Edge other) const{
    return vertex1 == other.vertex1 && vertex2 == other.vertex2;
}

File: Semester 2/Algoritmica Grafelor/Labs/Lab 2/graph.py
from exceptions import *
from random import randrange
from copy import deepcopy


def read_file(file_path):
    file = open(file_path, "r")
    n, m = map(int, file.readline().split())
    g = UndirectedGraph(n)
    for _ in range(m):
        vertex1, vertex2, edge_cost = map(int, file.readline().split())
        g.add_edge(vertex1, vertex2, edge_cost)
    file.close()
    return g


def write_file(file_path, g):
    file = open(file_path, "w")
    file.write("{0} {1}\n".format(g.count_vertices(), g.count_edges()))
    for node in g.vertices_iterator():
        for neighbour in g.neighbours_iterator(node):
            file.write("{0} {1} {2}\n".format(node, neighbour, g.get_edge_cost(node, neighbour)))
    file.close()


def random_graph(vertices_no, edges_no):
    g = UndirectedGraph()
    for i in range(vertices_no):
        g.add_vertex(i)
    for j in range(edges_no):
        vertex1 = randrange(vertices_no)
        vertex2 = randrange(vertices_no)
        while g.is_edge(vertex1, vertex2):
            vertex1 = randrange(vertices_no)
            vertex2 = randrange(vertices_no)
        g.add_edge(vertex1, vertex2, randrange(1000000))


class UndirectedGraph:
    def __init__(self, n=0, m=0):
        self._vertices = set()
        self._neighbours = dict()
        self._cost = dict()
        for i in range(n):
            self.add_vertex(i)
        for j in range(m):
            vertex1 = randrange(n)
            vertex2 = randrange(n)
            while self.is_edge(vertex1, vertex2):
                vertex1 = randrange(n)
                vertex2 = randrange(n)
            self.add_edge(vertex1, vertex2, randrange(1000000))

    def vertices_iterator(self):
        """
        Returns an iterator to the set of vertices.
        """
        for vertex in self._vertices:
            yield vertex

    def neighbours_iterator(self, vertex):
        """
        Returns an iterator to the set of (outbound) neighbours of a vertex.
        """
        if not self.is_vertex(vertex):
            raise VertexError("Invalid vertex.")
        for neighbour in self._neighbours[vertex]:
            yield neighbour

    def edges_iterator(self):
        """
        Returns an iterator to the set of edges.
        """
        for key, value in self._cost.items():
            yield key[0], key[1], value

    def is_vertex(self, vertex):
        """
        Returns True if vertex belongs to the graph.
        """
        return vertex in self._vertices

    def is_edge(self, vertex1, vertex2):
        """
        Returns True if the edge from vertex1 to vertex2 belongs to the graph.
        """
        if vertex1 > vertex2:
            vertex1, vertex2 = vertex2, vertex1
        return vertex1 in self._neighbours and vertex2 in self._neighbours[vertex1]

    def count_vertices(self):
        """
        Returns the number of vertices in the graph.
        """
        return len(self._vertices)

    def count_edges(self):
        """
        Returns the number of edges in the graph.
        """
        return len(self._cost)

    def degree(self, vertex):
        """
        Returns the number of edges with the start point vertex.
        """
        if vertex not in self._neighbours:
            raise VertexError("The specified vertex does not exist.")
        return len(self._neighbours[vertex])

    def get_edge_cost(self, vertex1, vertex2):
        """
        Returns the cost of an edge if it exists.
        """
        if vertex1 > vertex2:
            vertex1, vertex2 = vertex2, vertex1
        if (vertex1, vertex2) not in self._cost:
            raise EdgeError("The specified edge does not exist.")
        return self._cost[(vertex1, vertex2)]

    def set_edge_cost(self, vertex1, vertex2, new_cost):
        """
        Sets the cost of an edge in the graph if it exists.
        """
        if vertex1 > vertex2:
            vertex1, vertex2 = vertex2, vertex1
        if (vertex1, vertex2) not in self._cost:
            raise EdgeError("The specified edge does not exist.")
        self._cost[(vertex1, vertex2)] = new_cost

    def add_vertex(self, vertex):
        """
        Adds a vertex to the graph.
        """
        if self.is_vertex(vertex):
            raise VertexError("Cannot add a vertex which already exists.")
        self._vertices.add(vertex)
        self._neighbours[vertex] = set()

    def add_edge(self, vertex1, vertex2, edge_cost=0):
        """
        Adds an edge to the graph.
        """
        if vertex1 > vertex2:
            vertex1, vertex2 = vertex2, vertex1
        if self.is_edge(vertex1, vertex2):
            raise EdgeError("The specified edge already exists")
        if not self.is_vertex(vertex1) or not self.is_vertex(vertex2):
            raise EdgeError("The vertices on the edge do not exist.")
        self._neighbours[vertex1].add(vertex2)
        self._neighbours[vertex2].add(vertex1)
        self._cost[(vertex1, vertex2)] = edge_cost

    def remove_edge(self, vertex1, vertex2):
        """
        Removes an edge from the graph.
        """
        if vertex1 > vertex2:
            vertex1, vertex2 = vertex2, vertex1
        if not self.is_edge(vertex1, vertex2):
            raise EdgeError("The specified edge does not exist.")
        del self._cost[(vertex1, vertex2)]
        self._neighbours[vertex1].remove(vertex2)
        self._neighbours[vertex2].remove(vertex1)

    def remove_vertex(self, vertex):
        """
        Removes a vertex from the graph.
        """
        if not self.is_vertex(vertex):
            raise VertexError("Cannot remove a vertex which doesn't exist.")
        to_remove = []
        for node in self._neighbours[vertex]:
            to_remove.append(node)
        for node in to_remove:
            self.remove_edge(vertex, node)
        del self._neighbours[vertex]
        self._vertices.remove(vertex)

    def copy(self):
        """
        Returns a deepcopy of the graph.
        """
        return deepcopy(self)


class DirectedGraph:
    def __init__(self, n=0, m=0):
        self._vertices = set()
        self._neighbours = dict()
        self._transpose = dict()
        self._cost = dict()
        for i in range(n):
            self.add_vertex(i)
        for j in range(m):
            vertex1 = randrange(n)
            vertex2 = randrange(n)
            while self.is_edge(vertex1, vertex2):
                vertex1 = randrange(n)
                vertex2 = randrange(n)
            self.add_edge(vertex1, vertex2, randrange(1000000))

    def vertices_iterator(self):
        """
        Returns an iterator to the set of vertices.
        """
        for vertex in self._vertices:
            yield vertex

    def neighbours_iterator(self, vertex):
        """
        Returns an iterator to the set of (outbound) neighbours of a vertex.
        """
        if not self.is_vertex(vertex):
            raise VertexError("Invalid vertex.")
        for neighbour in self._neighbours[vertex]:
            yield neighbour

    def transpose_iterator(self, vertex):
        """
        Returns an iterator to the set of (inbound) neighbours of a vertex.
        """
        if not self.is_vertex(vertex):
            raise VertexError("Invalid vertex.")
        for neighbour in self._transpose[vertex]:
            yield neighbour

    def edges_iterator(self):
        """
        Returns an iterator to the set of edges.
        """
        for key, value in self._cost.items():
            yield key[0], key[1], value

    def is_vertex(self, vertex):
        """
        Returns True if vertex belongs to the graph.
        """
        return vertex in self._vertices

    def is_edge(self, vertex1, vertex2):
        """
        Returns True if the edge from vertex1 to vertex2 belongs to the graph.
        """
        return vertex1 in self._neighbours and vertex2 in self._neighbours[vertex1]

    def count_vertices(self):
        """
        Returns the number of vertices in the graph.
        """
        return len(self._vertices)

    def count_edges(self):
        """
        Returns the number of edges in the graph.
        """
        return len(self._cost)

    def in_degree(self, vertex):
        """
        Returns the number of edges with the endpoint vertex.
        """
        if vertex not in self._transpose:
            raise VertexError("The specified vertex does not exist.")
        return len(self._transpose[vertex])

    def out_degree(self, vertex):
        """
        Returns the number of edges with the start point vertex.
        """
        if vertex not in self._neighbours:
            raise VertexError("The specified vertex does not exist.")
        return len(self._neighbours[vertex])

    def get_edge_cost(self, vertex1, vertex2):
        """
        Returns the cost of an edge if it exists.
        """
        if (vertex1, vertex2) not in self._cost:
            raise EdgeError("The specified edge does not exist.")
        return self._cost[(vertex1, vertex2)]

    def set_edge_cost(self, vertex1, vertex2, new_cost):
        """
        Sets the cost of an edge in the graph if it exists.
        """
        if (vertex1, vertex2) not in self._cost:
            raise EdgeError("The specified edge does not exist.")
        self._cost[(vertex1, vertex2)] = new_cost

    def add_vertex(self, vertex):
        """
        Adds a vertex to the graph.
        """
        if self.is_vertex(vertex):
            raise VertexError("Cannot add a vertex which already exists.")
        self._vertices.add(vertex)
        self._neighbours[vertex] = set()
        self._transpose[vertex] = set()

    def add_edge(self, vertex1, vertex2, edge_cost=0):
        """
        Adds an edge to the graph.
        """
        if self.is_edge(vertex1, vertex2):
            raise EdgeError("The specified edge already exists")
        if not self.is_vertex(vertex1) or not self.is_vertex(vertex2):
            raise EdgeError("The vertices on the edge do not exist.")
        self._neighbours[vertex1].add(vertex2)
        self._transpose[vertex2].add(vertex1)
        self._cost[(vertex1, vertex2)] = edge_cost

    def remove_edge(self, vertex1, vertex2):
        """
        Removes an edge from the graph.
        """
        if not self.is_edge(vertex1, vertex2):
            raise EdgeError("The specified edge does not exist.")
        del self._cost[(vertex1, vertex2)]
        self._neighbours[vertex1].remove(vertex2)
        self._transpose[vertex2].remove(vertex1)

    def remove_vertex(self, vertex):
        """
        Removes a vertex from the graph.
        """
        if not self.is_vertex(vertex):
            raise VertexError("Cannot remove a vertex which doesn't exist.")
        to_remove = []
        for node in self._neighbours[vertex]:
            to_remove.append(node)
        for node in to_remove:
            self.remove_edge(vertex, node)
        to_remove = []
        for node in self._transpose[vertex]:
            to_remove.append(node)
        for node in to_remove:
            self.remove_edge(node, vertex)
        del self._neighbours[vertex]
        del self._transpose[vertex]
        self._vertices.remove(vertex)

    def copy(self):
        """
        Returns a deepcopy of the graph.
        """
        return deepcopy(self)
